# Fine tuning classification example

We will fine-tune a `babbage-002` classifier (replacement for the `ada` models) to distinguish between the two sports: Baseball and Hockey.

**Fine-tuning classification refers to the process of customizing a pre-trained model (like GPT or other machine learning models) to perform well on a specific classification task. In this context, the goal is to train the model to categorize inputs (e.g., text, images, or other data) into predefined labels or classes.**

# **Understanding Fine-Tuning for Classification**

Pre-trained Models: Models like GPT and BERT are trained on large datasets to learn general patterns like language structure and relationships.

Fine-Tuning: Fine-tuning specializes a pre-trained model by training it further on a smaller, task-specific dataset to adapt to the desired task.

**Classification Tasks:**

Classification tasks involve categorizing data into specific classes. Examples include:
        
Text Classification: Classifying emails as spam or not spam.
        
Sentiment Analysis: Determining whether a review is positive, negative, or neutral.
        
Topic Detection: Identifying the topic of an article (e.g., sports, technology, politics).

**Email Spam Classification**

Task: Categorize emails as either Spam or Not Spam.
    
Example:
        
Input: "Congratulations! You've won a $1,000 gift card. Click the link to claim your prize!"
        
Output: "Spam"

2.**Sentiment Analysis**

Task: Classify the sentiment of a review or comment as Positive, Negative, or Neutral.
    
Example:
        
Input: "The product quality is excellent, and I’m very satisfied with the purchase!"
        
Output: "Positive"

In [4]:
!pip install --quiet openai

In [5]:
!pip install --upgrade openai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 389.5/389.5 kB 8.9 MB/s eta 0:00:00
  Attempting uninstall: openai
    Found existing installation: openai 1.54.4
    Uninstalling openai-1.54.4:
      Successfully uninstalled openai-1.54.4


In [6]:
from sklearn.datasets import fetch_20newsgroups
import pandas as pd
import openai
import os

#client = openai.OpenAI(api_key=os.environ.get("OPENAI_API_KEY", "<your OpenAI API key if not set as env var>"))

categories = ['rec.sport.baseball', 'rec.sport.hockey']
sports_dataset = fetch_20newsgroups(subset='train', shuffle=True, random_state=42, categories=categories)

 ## Data exploration
 The newsgroup dataset can be loaded using sklearn. First we will look at the data itself:

In [7]:
print(sports_dataset['data'][0])

From: dougb@comm.mot.com (Doug Bank)
Subject: Re: Info needed for Cleveland tickets
Reply-To: dougb@ecs.comm.mot.com
Organization: Motorola Land Mobile Products Sector
Distribution: usa
Nntp-Posting-Host: 145.1.146.35
Lines: 17

In article <1993Apr1.234031.4950@leland.Stanford.EDU>, bohnert@leland.Stanford.EDU (matthew bohnert) writes:

|> I'm going to be in Cleveland Thursday, April 15 to Sunday, April 18.
|> Does anybody know if the Tribe will be in town on those dates, and
|> if so, who're they playing and if tickets are available?

The tribe will be in town from April 16 to the 19th.
There are ALWAYS tickets available! (Though they are playing Toronto,
and many Toronto fans make the trip to Cleveland as it is easier to
get tickets in Cleveland than in Toronto.  Either way, I seriously
doubt they will sell out until the end of the season.)

-- 
Doug Bank                       Private Systems Division
dougb@ecs.comm.mot.com          Motorola Communications Sector
dougb@nwu.edu       

In [9]:
sports_dataset['target_names']

['rec.sport.baseball', 'rec.sport.hockey']

In [10]:
sports_dataset.target_names[sports_dataset['target'][0]]


'rec.sport.baseball'

In [11]:
len_all, len_baseball, len_hockey = len(sports_dataset.data), len([e for e in sports_dataset.target if e == 0]), len([e for e in sports_dataset.target if e == 1])
print(f"Total examples: {len_all}, Baseball examples: {len_baseball}, Hockey examples: {len_hockey}")

Total examples: 1197, Baseball examples: 597, Hockey examples: 600


One sample from the baseball category can be seen above. It is an email to a mailing list. We can observe that we have 1197 examples in total, which are evenly split between the two sports.

## Data Preparation
We transform the dataset into a pandas dataframe, with a column for prompt and completion. The prompt contains the email from the mailing list, and the completion is a name of the sport, either hockey or baseball. For demonstration purposes only and speed of fine-tuning we take only 300 examples. In a real use case the more examples the better the performance.

In [12]:
import pandas as pd

labels = [sports_dataset.target_names[x].split('.')[-1] for x in sports_dataset['target']]
texts = [text.strip() for text in sports_dataset['data']]
df = pd.DataFrame(zip(texts, labels), columns = ['prompt','completion']) #[:300]
df.head()

prompt completion
0  From: dougb@comm.mot.com (Doug Bank)\nSubject:...   baseball
1  From: gld@cunixb.cc.columbia.edu (Gary L Dare)...     hockey
2  From: rudy@netcom.com (Rudy Wade)\nSubject: Re...   baseball
3  From: monack@helium.gas.uug.arizona.edu (david...     hockey
4  Subject: Let it be Known\nFrom: <ISSBTL@BYUVM....   baseball

Both baseball and hockey are single tokens. We save the dataset as a jsonl file.

In [13]:
df.to_json("sport2.jsonl", orient='records', lines=True)

### Data Preparation tool
We can now use a data preparation tool which will suggest a few improvements to our dataset before fine-tuning. Before launching the tool we update the openai library to ensure we're using the latest data preparation tool. We additionally specify `-q` which auto-accepts all suggestions.

In [16]:
!openai tools fine_tunes.prepare_data -f sport2.jsonl -q

Analyzing...

- Your file contains 1197 prompt-completion pairs
- Based on your data it seems like you're trying to fine-tune a model for classification
- For classification, we recommend you try one of the faster and cheaper models, such as `ada`
- For classification, you can estimate the expected model performance by keeping a held out dataset, which is not used for training
- There are 11 examples that are very long. These are rows: [134, 200, 281, 320, 404, 595, 704, 838, 1113, 1139, 1174]
For conditional generation, and for classification the examples shouldn't be longer than 2048 tokens.
- Your data does not contain a common separator at the end of your prompts. Having a separator string appended to the end of the prompt makes it clearer to the fine-tuned model where the completion should begin. See https://platform.openai.com/docs/guides/fine-tuning/preparing-your-dataset for more detail and examples. If you intend to do open-ended generation, then you should leave the prompts e

The tool helpfully suggests a few improvements to the dataset and splits the dataset into training and validation set.

A suffix between a prompt and a completion is necessary to tell the model that the input text has stopped, and that it now needs to predict the class. Since we use the same separator in each example, the model is able to learn that it is meant to predict either baseball or hockey following the separator.
A whitespace prefix in completions is useful, as most word tokens are tokenized with a space prefix.
The tool also recognized that this is likely a classification task, so it suggested to split the dataset into training and validation datasets. This will allow us to easily measure expected performance on new data.

## Fine-tuning
The tool suggests we run the following command to train the dataset. Since this is a classification task, we would like to know what the generalization performance on the provided validation set is for our classification use case.

We can simply copy the suggested command from the CLI tool. We specifically add `-m ada` to fine-tune a cheaper and faster ada model, which is usually comperable in performance to slower and more expensive models on classification use cases.

In [17]:
ls

 sample_data/  'sport2_prepared_train (1).jsonl'  'sport2_prepared_valid (1).jsonl'
 sport2.jsonl   sport2_prepared_train.jsonl        sport2_prepared_valid.jsonl


In [ ]:
import openai

# Set API key
openai.api_key = "your_actual_api_key"

The model is successfully trained in about ten minutes. You can watch the finetune happen on [https://platform.openai.com/finetune/](https://platform.openai.com/finetune/)

You can also check on its status programatically:

In [ ]:
# Upload training and validation files
train_file = client.files.create(file=open("sport2_prepared_train.jsonl", "rb"), purpose="fine-tune")
valid_file = client.files.create(file=open("sport2_prepared_valid.jsonl", "rb"), purpose="fine-tune")

# Start fine-tuning job
fine_tuning_job = client.fine_tuning.jobs.create(training_file=train_file.id, validation_file=valid_file.id, model="babbage-002")

print(fine_tuning_job)

In [ ]:
fine_tune_results = client.fine_tuning.jobs.retrieve(fine_tuning_job.id)
print(fine_tune_results.finished_at)

1704414393


### [Advanced] Results and expected model performance
We can now download the results file to observe the expected performance on a held out validation set.

In [ ]:
fine_tune_results = client.fine_tuning.jobs.retrieve(fine_tuning_job.id).result_files
result_file = client.files.retrieve(fine_tune_results[0])
content = client.files.content(result_file.id)
# save content to file
with open("result.csv", "wb") as f:
    f.write(content.text.encode("utf-8"))

In [ ]:
results = pd.read_csv('result.csv')
results[results['train_accuracy'].notnull()].tail(1)

step  train_loss  train_accuracy  valid_loss  valid_mean_token_accuracy
2843  2844         0.0             1.0         NaN                        NaN

The accuracy reaches 99.6%. On the plot below we can see how accuracy on the validation set increases during the training run.

In [ ]:
results[results['train_accuracy'].notnull()]['train_accuracy'].plot()

## Using the model
We can now call the model to get the predictions.

In [ ]:
test = pd.read_json('sport2_prepared_valid.jsonl', lines=True)
test.head()

prompt completion
0  From: gld@cunixb.cc.columbia.edu (Gary L Dare)...     hockey
1  From: smorris@venus.lerc.nasa.gov (Ron Morris ...     hockey
2  From: golchowy@alchemy.chem.utoronto.ca (Geral...     hockey
3  From: krattige@hpcc01.corp.hp.com (Kim Krattig...   baseball
4  From: warped@cs.montana.edu (Doug Dolven)\nSub...   baseball

We need to use the same separator following the prompt which we used during fine-tuning. In this case it is `\n\n###\n\n`. Since we're concerned with classification, we want the temperature to be as low as possible, and we only require one token completion to determine the prediction of the model.

In [ ]:
ft_model = fine_tune_results.fine_tuned_model

# note that this calls the legacy completions api - https://platform.openai.com/docs/api-reference/completions
res = client.completions.create(model=ft_model, prompt=test['prompt'][0] + '\n\n###\n\n', max_tokens=1, temperature=0)
res.choices[0].text


' hockey'

To get the log probabilities, we can specify logprobs parameter on the completion request

In [ ]:
res = client.completions.create(model=ft_model, prompt=test['prompt'][0] + '\n\n###\n\n', max_tokens=1, temperature=0, logprobs=2)
res.choices[0].logprobs.top_logprobs

[{' hockey': 0.0, ' Hockey': -22.504879}]

We can see that the model predicts hockey as a lot more likely than baseball, which is the correct prediction. By requesting log_probs, we can see the prediction (log) probability for each class.

### Generalization
Interestingly, our fine-tuned classifier is quite versatile. Despite being trained on emails to different mailing lists, it also successfully predicts tweets.

In [ ]:
sample_hockey_tweet = """Thank you to the
@Canes
 and all you amazing Caniacs that have been so supportive! You guys are some of the best fans in the NHL without a doubt! Really excited to start this new chapter in my career with the
@DetroitRedWings
 !!"""
res = client.completions.create(model=ft_model, prompt=sample_hockey_tweet + '\n\n###\n\n', max_tokens=1, temperature=0, logprobs=2)
res.choices[0].text

' hockey'

In [ ]:
sample_baseball_tweet="""BREAKING: The Tampa Bay Rays are finalizing a deal to acquire slugger Nelson Cruz from the Minnesota Twins, sources tell ESPN."""
res = client.completions.create(model=ft_model, prompt=sample_baseball_tweet + '\n\n###\n\n', max_tokens=1, temperature=0, logprobs=2)
res.choices[0].text

# **NOTE: HOW IT IS CLASSIFIER WHERE IT A HOCKEY OR BASEBALL.**

1. **Keywords and Terms**

    The tweet includes baseball-specific terms like:
        "slugger" (a term commonly used in baseball for a powerful hitter).
        Names of MLB teams: "Tampa Bay Rays" and "Minnesota Twins".
    These terms strongly associate with baseball, as they are part of its unique terminology.

2. **Lack of Hockey-Specific Terms**

    The tweet does not mention any hockey-specific elements such as:
        NHL team names.
        Hockey-related terminology like "puck", "goal", or "NHL".
    The absence of hockey-related terms helps the model exclude Hockey as a label.

3. **Pre-trained Patterns from Fine-Tuning**

    During fine-tuning, the model learns the association of teams like Tampa Bay Rays and Minnesota Twins with baseball.
    Similarly, it learns that phrases like "acquire slugger" commonly occur in the context of baseball trades, not hockey.

4. **Contextual Knowledge**

    Baseball involves frequent trades and acquisitions of players, especially with terms like "slugger" or "deal".
    The model leverages its general knowledge to understand that this scenario fits the context of baseball.

5. **Conclusion**

The classifier correctly labels the tweet as Baseball based on the presence of team names, terminology, and context learned during fine-tuning. The absence of hockey-related clues further reinforces the choice.
